src/ORM/DB.php

```diff
@@ -696,4 +704,100 @@ public static function alteration_message($message, $type = "")
+        // is forced as ready without having to check all the superclasses.
+        if (!empty(self::$databaseReadyClasses[$class])) {
+            return true;
+        }
+
+        // Check if all tables and fields required for the class exist in the database.
+        $requiredClasses = ClassInfo::dataClassesFor($class);
+        $schema = DataObject::getSchema();
+        foreach ($requiredClasses as $required) {
+            // Skip test classes, as not all test classes are scaffolded at once
+            if (is_a($required, TestOnly::class, true)) {
+                continue;
+            }
+
+            // Don't check again if we already know the db is ready for this class.
+            if (!empty(self::$databaseReadyClasses[$class])) {
+                continue;
+            }
+
+            // if any of the tables aren't created in the database
+            $table = $schema->tableName($required);
+            if (!ClassInfo::hasTable($table)) {
+                return false;
+            }
+
+            // HACK: DataExtensions aren't applied until a class is instantiated for
+            // the first time, so create an instance here.
+            singleton($required);
+
+            // if any of the tables don't have all fields mapped as table columns
+            $dbFields = DB::field_list($table);
+            if (!$dbFields) {
+                return false;
+            }
+
+            $objFields = $schema->databaseFields($required, false);
+            $missingFields = array_diff_key($objFields, $dbFields);
+
+            if ($missingFields) {
+                return false;
+            }
+
+            // Add each ready class to the cached array.
+            self::$databaseReadyClasses[$required] = true;
+        }
+
+        return true;
+    }
+
+    /**
+     * Resets the databaseReadyClasses cache.
+     *
+     * @param string|null $class The specific class to be cleared.
+     * If not passed, the cache for all classes is cleared.
+     * @param bool $clearFullHeirarchy Whether to clear the full class hierarchy or only the given class.
+     */
+    public static function clearDatabaseIsReady(?string $class = null, bool $clearFullHierarchy = true)
+    {
+        if ($class) {
+            $clearClasses = [$class];
+            if ($clearFullHierarchy) {
+                $clearClasses = ClassInfo::dataClassesFor($class);
+            }
+            foreach ($clearClasses as $clear) {
+                unset(self::$databaseReadyClasses[$clear]);
+            }
+        } else {
+            self::$databaseReadyClasses = [];
+        }
+    }
+
+    /**
+     * For the databaseIsReady call to return a certain value for the given class - used for testing
+     *
+     * @param string $class The class to be forced as ready/not ready.
+     * @param boolean $isReady The value to force.
+     */
+    public static function forceDatabaseIsReady(string $class, bool $isReady)
```

Comment:
This is method is pretty unecessary IMO and it just clutters up the API.  I think you're better off just update the visibility of the property to public via Reflection in the unit test and then updating the property directly

Reply:
`Security::force_database_is_ready()` is a public method, which is what's called by the test. I agree that ideally the method wouldn't exist in the new location... should I replace it with reflection in `Security::force_database_is_ready()` (so that method still does what it has always done)?
And then any new tests that might need that functionality can use reflection directly in the tests?

Reply:
Leave the existing `Security` method as is and use reflection within the new unit test. That will minimise the amount of API changes

Reply:
nvm I realised a much cleaner way to handle this, which goes in line with https://github.com/silverstripe/silverstripe-framework/pull/10041#discussion_r843440070